shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCXConnection.scala

```diff
@@ -124,46 +230,72 @@ class UCXConnection(peerExecutorId: Int, ucx: UCX) extends Connection with Loggi
    * 2) then comes the type, which gets 4 bits
    * 3) the remaining 32 bits are used for buffer specific tags
    */
-  private val requestMsgType:  Long = 0x00000000000000000L
-  private val responseMsgType: Long = 0x0000000000000000AL
   private val bufferMsgType:   Long = 0x0000000000000000BL
 
-  override def composeRequestTag(requestType: RequestType.Value): Long = {
-    val requestTypeId = requestType match {
-      case RequestType.MetadataRequest => 0
-      case RequestType.TransferRequest => 1
-    }
-    requestMsgType | requestTypeId
-  }
+  // message type mask for UCX tags. The only message type using tags
+  // is `bufferMsgType`
+  val msgTypeMask    = 0x000000000000000FL
 
-  protected def composeResponseTag(peerClientId: Long, bufferTag: Long): Long = {
-    // response tags are [peerClientId, 1, bufferTag]
-    composeTag(composeUpperBits(peerClientId, responseMsgType), bufferTag)
-  }
+  // UCX Active Message masks (we can use up to 5 bits for these ids)
+  val amRequestMask  = 0x0000000F
+  val amResponseMask = 0x0000001F
+
+  // We pick the 5th bit set to 1 as a "response" active message
+  val amResponseFlag = 0x00000010
+
+  // pick up the lower and upper parts of a long
+  val lowerBitsMask  = 0x00000000FFFFFFFFL
+  val upperBitsMask  = 0xFFFFFFFF00000000L
 
   protected def composeBufferTag(peerClientId: Long, bufferTag: Long): Long = {
     // buffer tags are [peerClientId, 0, bufferTag]
     composeTag(composeUpperBits(peerClientId, bufferMsgType), bufferTag)
   }
 
-  private def composeTag(upperBits: Long, lowerBits: Long): Long = {
-    if ((upperBits & 0xFFFFFFFF00000000L) != upperBits) {
+  protected def composeRequestAmId(requestType: RequestType.Value): Int = {
+    val amId = requestType.id
+    if ((amId & amRequestMask) != amId) {
       throw new IllegalArgumentException(
-        s"Invalid tag, upperBits would alias: ${TransportUtils.formatTag(upperBits)}")
+        s"Invalid request amId, it must be 4 bits: ${TransportUtils.toHex(amId)}")
+    }
+    amId
+  }
+
+  protected def composeRequestHeader(executorId: Long, txId: Long): Long = {
+    composeTag(executorId << 32, txId)
+  }
+
+  def extractExecutorId(header: Long): Long = {
+    (header & upperBitsMask) >> 32
```

Comment:
I made some changes to this to shift-then-mask (https://github.com/NVIDIA/spark-rapids/pull/2409/commits/42397f8e0101ee389c6667b658bfa35d5f77eb2a). I also added some requires around the sign of executorId, since it better not be negative.